### PR TITLE
implements best legend place holder

### PR DESCRIPTION
**Pull request by lmiq**
Implemented the best legend position placeholder: It will choose between bottomleft/right or topleft/right, depending on a sample of the data, trying to maximize the difference to the closest point. The function is fast (5-10 microseconds), and should be good enough for a guess of the best position. 

Tries to address: https://github.com/JuliaPlots/Plots.jl/issues/3399

Test it with:
```julia
using Plots
scatter(randn(100), randn(100))
```

### CODE REVIEW

**Comment by t-bltg** [comment]
Nice, I'm going to try it now.

Don't forget to run `using JuliaFormatter; format(["src", "test"])` for the check to pass.

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1155]:
```suggestion
    xlims = xlims(plt)
    ylims = ylims(plt)
```

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1141]: I think we can unroll by hand and avoid the allocations here ?

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1160]:
```suggestion
    for (i, lim) in enumerate(Iterators.product(xlims,ylims))

```

File: src/utils.jl
Comment (lmiq) [reply src/utils.jl:1141]: I don't think the allocations are coming from there, `lim` is a tuple, known at compile time. I've tested that. I think the allocations are coming from the fact that `xlims` are returned as `Tuple{Any,Any}`, which is beyond what I can deal with in this functions. I tried adding a function barrier but that removed 2 allocations only and was not worth the complication (the effect on time was irrelevant).

**Comment by t-bltg** [comment]
Here are the changed reference images.
`ref4`
ref4
`ref7`
ref7
`ref8`
ref8
`ref12`
ref12
`ref13`
ref13
`ref15`
ref15
`ref17`
ref17
`ref18`
ref18
`ref19`
ref19
`ref23`
ref23
`ref25`
ref25
`ref26`
ref26
`ref29`
ref29
`ref30`
ref30
`ref36`
ref36
`ref37`
ref37
`ref44`
ref44
`ref46`
ref46
`ref48`
ref48
`ref51`
ref51
`ref53`
ref53
`ref56`
ref56
`ref58`
ref58
`ref61`
ref61
`ref62`
![ref62](https://user-images.githubusercontent.com/13423344/203096381-1e89230c-5add-43ca-

File: src/backends/gr.jl
Comment (t-bltg) [inline src/backends/gr.jl:1115]: What I meant was:

```suggestion
    leg_str = string(_guess_best_legend_position(lp, sp))
```

and remove `    if (leg_str = string(lp)) == "best"`.

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1156]:
```suggestion
    lp === :best || return lp
```

Nitpick on concision.

**Comment by lmiq** [comment]
> Here are the changed reference images (25 out of 65).

Actually most of them look weird to me, it is not what I was expecting. But now that I know those tests exist, I can check each of them before committing changes. 

(The histogram is case is a case apart, as the plot does not show the data series, and what I do know is based on the data series).

**Comment by t-bltg** [comment]
> But now that I know those tests exist, I can check each of them before committing changes.

The refs files are checked when running `test Plots`.

You can set `julia> ENV["VISUAL_REGRESSION_TESTS_AUTO"] = true` before `pkg> test Plots` (see https://github.com/JuliaPlots/Plots.jl/blob/e6a068934f3f0a4a3b6e0a8dc6ca81744be81717/test/runtests.jl#L53) in order to update them automatically, and partially run the test-suite (faster).

**Comment by lmiq** [comment]
I'll close the PR for now to work on it further without disturbing you.

**Comment by t-bltg** [comment]
You could also have marked it as a draft (it disables notifications ;)).

**Comment by lmiq** [comment]
I think now it is doing mostly a good job. It could be slightly better if I was able to get the `:size` attribute of the plot there, but I'm getting an error trying to access `plt.attr.explicit[:size]` from within `_guess_best_legend_position`. 

Nevertheless, from my tests, the default positions now in the tests look mostly good, and better than before.

**Comment by t-bltg** [comment]
> but I'm getting an error trying to access plt.attr.explicit[:size] from within _guess_best_legend_position

I think `plt[:size]` would work.

**Comment by t-bltg** [comment]
Really nice work. I only think `ref19` would be better with a top-right margin, but this is marginal.

A test is also broken:
```julia
Examples: Error During Test at [...]/Plots.jl/test/test_backends.jl:211
  Got exception outside of a @test
  MethodError: no method matching typemax(::Type{Real})
  Closest candidates are:
    typemax(::Union{DateTime, Type{DateTime}}) at [...]/share/julia/stdlib/v1.8/Dates/src/types.jl:453
    typemax(::Union{Date, Type{Date}}) at [...]/share/julia/stdlib/v1.8/Dates/src/types.jl:455
    typemax(::Union{Time, Type{Time}}) at [...]/share/julia/stdlib/v1.8/Dates/src/types.jl:457
    ...
  Stacktrace:
    [1] (::Plots.var"#82#84"{DataType})(#unused#::Int64)
      @ Plots [...]/Plots.jl/src/utils.jl:1161
    [2] ntuple(f::Plots.var"#82#84"{DataType}, n::Int64)
      @ Base ./ntuple.jl:19
    [3] _guess_best_legend_position(lp::Symbol, plt::Plots.Subplot{Plots.GRBackend}; nsamples::Int64)
      @ Plots [...]/Plots.jl/src/utils.jl:1161
    [4] _guess_best_legend_position
      @ [...]/Plots.jl/src/utils.jl:1155 [inlined]
    [5] gr_legend_pos(sp::Plots.Subplot{Plots.GRBackend}, leg::NamedTuple{(:yoffset, :xoffset, :base_markersize, :base_factor, :has_title, :vertical, :entries, :space, :texth, :textw, :span, :pad, :dy, :dx, :w, :h), Tuple{Float64, Float64, Float64, Float64, Bool, Bool, Int64, Float64, Float64, Float64, Float64, Float64, Float64, Float64, Float64, Float64}}, vp::Plots.GRViewport{Float64})
      @ Plots [...]/Plots.jl/src/backends/gr.jl:1120
    [6] gr_add_legend(sp::Plots.Subplot{Plots.GRBackend}, leg::NamedTuple{(:yoffset, :xoffset, :base_markersize, :base_factor, :has_title, :vertical, :entries, :space, :texth, :textw, :span, :pad, :dy, :dx, :w, :h), Tuple{Float64, Float64, Float64, Float64, Bool, Bool, Int64, Float64, Float64, Float64, Float64, Float64, Float64, Float64, Float64, Float64}}, viewport_area::Plots.GRViewport{Float64})
      @ Plots [...]/Plots.jl/src/backends/gr.jl:991
```

**Comment by lmiq** [comment]
Let me fix all those things and see if I can use the plot size to improve corner case (elongated plots). When I'm ready I remove the Draft flag (sorry not making it Draft before, I couldn't find where to do it until you commented it).

**Comment by lmiq** [comment]
@t-bltg 

Concerning example 19: it is possible to improve the position of the guess by increasing the `nsamples` parameter. But that comes with a cost. 

Maybe we could have two options: `:auto`, with the current behavior, and `:best`, in which the complete data is sampled to place the legend (slower for large data sets).  What you think?

**Comment by t-bltg** [comment]
I think the current is still acceptable for ref19.

Can you post the results of a simple benchmark: `@btime show(devnull, plot(1:2))` with an without this PR, to see what the induced time increase looks like ?
If it is marginal, we can maybe increase the default `nsamples`.

I'd be in favor to `:best` as default, but maybe @BeastyBlacksmith has another opinion.
PyPlot's default is `:best` I think.

File: src/backends/gr.jl
Comment (t-bltg) [inline src/backends/gr.jl:1107]: This can be simplified to 
```julia
leg_str = string(_guess_best_legend_position(lp, sp))
```

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1167]: Would it make sense to fix the type instability of `xlims`, `ylims` ?

**Comment by t-bltg** [comment]
Nice, I really like where this is going.

I don't have anything to add (PR is tested, benchmarked, carefully implemented ...).

When you think that this `PR` is ready, you can run `Plots` test suite locally, update the reference images, and submit a `PR` to https://github.com/JuliaPlots/PlotReferenceImages.jl.

If we are satisfied with the new refs images (changes can be inspected in the github interface), we merge it, and we run the `Plots` test suite here.

When `ci` is green and `PR` approved here -> merge.

File: test/test_utils.jl
Comment (t-bltg) [inline test/test_utils.jl:198]:
```suggestion
    @test Plots._guess_best_legend_position(:best, plt) === :topleft
```

And :arrow_down: as well.

**Comment by lmiq** [comment]
Corresponding image PR: https://github.com/JuliaPlots/PlotReferenceImages.jl/pull/142

**Review by t-bltg — APPROVED** [review]
I'd like @BeastyBlacksmith to comment before merging the reference images.

To me, there is no ambiguity on the automatic placement (at least on the ref images).

**Comment by lmiq** [comment]
I just updated a comment. Nothing will change.

**Review by BeastyBlacksmith — APPROVED** [review]
I'd be more happy if it kept `:topright` in the cases where it is impossible not to overlap with data (or maybe even choose `:outertopright` then).
But I like it either way. Good work!

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1164]:
```suggestion
    scale = (maximum(xl) - minimum(xl), maximum(yl) - minimum(yl))
```

Isn't it faster to use `extrema` here (one single pass) ?

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1182]:
```suggestion
    return _guess_best_legend_position(xlims(plt), ylims(plt), plt, nsamples)
```

File: src/utils.jl
Comment (lmiq) [reply src/utils.jl:1164]: Doesn't make any difference, and makes the code uglier, IMO (have do add two lines to first compute the extrema and then use it)

**Comment by lmiq** [comment]
> where it is impossible not to overlap

If we could obtain the coordinates of the legend box we could try that. But I didn't find how to do it, so we are just examining the plot edges relationship with the data. We can try some different heuristics, such as summing some distances, but we would need a case where the current strategy is clearly bad. Also biased examples against any other strategy should be tested.

File: src/utils.jl
Comment (t-bltg) [inline src/utils.jl:1213]: Couldn't we detect overlap here by adding a distance threshold ?
Such that we use `findfirst` instead and on failure (`nothing`) return `:topright` as fallback.

Comment (lmiq) [reply src/utils.jl:1213]: To actually detect overlap we need the coordinates of the legend box. With that information we could compute at roughly the same cost how many atoms fall into it. 

Using any other heuristic without that  may be of course fine in particular cases, but I'm not sure if it is worth the trouble.

Comment (t-bltg) [reply src/utils.jl:1213]: I dislike introducing heuristics in a deterministic algorithm too.
I'm happy with the current, but I'm also anticipating users complaining about the new behavior.
Let me think about that while ci is running.

Comment (lmiq) [reply src/utils.jl:1213]: I didn't mean something with aleatory behavior, just a different strategy.

I'm pretty sure someone will complain, but that will happen with any strategy :-D

What I can try rather easily is to use a sum of inverse distances, or similar, and see if you like the results better.

Comment (t-bltg) [reply src/utils.jl:1213]: Regarding the `coordinates of the legend box`, the legend geometry is computed here https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1248-L1265

And you namedtuple `leg` is accessible in the parent calling function https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1105, so you can maybe pass this nt to `_guess_best_legend_position` (and make it conditional, defaulting to `nothing` for other backends that might not compute legend geometry).

My little drawing in https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1219-L1224 might be useful here.

Comment (BeastyBlacksmith) [reply src/utils.jl:1213]: I'm against calling backend functions in frontend code. We'd need a utility function in the frontend then, that backends can overload.

**Comment by lmiq** [comment]
I'll post the results of this different placement strategy, which does better if there is a bunch of points in one corner, but the closest overlap is in the other (should I have created a new branch? I'm still confused with this)

**Comment by lmiq** [comment]
Here they are: https://github.com/JuliaPlots/PlotReferenceImages.jl/pull/143/files

It actually looks good, IMO. Maybe better, particularly in the scatter plots.

**Comment by t-bltg** [comment]
Thanks, merged, let's try that now.

**Comment by t-bltg** [comment]
I've have cleaned up reference images between versions `1.36.3` and `1.36.4`.

**Comment by t-bltg** [comment]
Should we merge ?

**Comment by t-bltg** [comment]
Thanks @lmiq, this is a really nice contribution :tada: .

**Pull request merged**
